src/gpuarray_reduction.c

```diff
@@ -0,0 +1,548 @@
+/* Includes */
+#ifdef _MSC_VER
+#define _CRT_SECURE_NO_WARNINGS
+#endif
+
+#include <assert.h>
+#include <stdarg.h>
+#include <stddef.h>
+#if !defined(_MSC_VER) || _MSC_VER < 1600
+#include <stdint.h>
+#endif
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+
+#include "private.h"
+#include "gpuarray/array.h"
+#include "gpuarray/error.h"
+#include "gpuarray/kernel.h"
+#include "gpuarray/util.h"
+
+#include "util/strb.h"
+
+
+/* Datatypes */
+struct GEN_KERNEL_CTX{
+	int         numIdx;
+	const int*  isReduced;
+	int         numRedIdx;
+	int         numFreeIdx;
+	const char* dstMaxType;
+	const char* dstArgmaxType;
+};
+typedef struct GEN_KERNEL_CTX GEN_KERNEL_CTX;
```

Comment:
@abergeron A prefix of `_` followed by a capital letter, or a prefix of `__`, are reserved by ISO C (and other offshoots of C like C++, OpenCL and CUDA) for the C implementation.

I propose to use `gen_kernel_ctx`.